docs/rholang/rholangmatchingtut.md

```diff
@@ -0,0 +1,181 @@
+
+Note further that since `P = P | Nil` for any process `P`, the pattern `@{x | y}` can match with a process with no parallel components. Likewise, `@Nil!(10 | 20)` would send successfully to something of the form `for( @{x | y | z} <- @Nil ){ Nil }`, where we've added an extra variable on the receiving pattern.
+
+Note that we can write patterns with parallel processes that are not just process variables, such as `@{ @Nil!(x) | for( z <- c ){ z!(Nil) } }`, etc.
+
+## Logical Connectives In Patterns
+
+Logical connectives can be incorporated into patterns, but we cannot send or receive over a channel which contains logical connectives. For example, it is incorrect to write
+
+    @{ 10 \/ 20 }!(Nil)
+
+We can, however, correctly write
+
+    for( @{@{ 10 \/ 20 }}!(z) <- @Nil ){ @Nil!(z) }
+
+which listens on `@Nil` for a process which sends either to `@10` or to `@20`. Whatever is being sent is grabbed and sent over `@Nil` in the body. If we use `/\` in a pattern, any free variables which are in the pattern will be bound to their corresponding parts. However, if we use `\/`, none of the variables in the parse tree below the node are bound. In particular, if `P1 \/ P2` is in a pattern, the program will fail to run if any free variables in `P1` or `P2` appear in the body, since they cannot be guaranteed to bind to anything.
+
+We note here that in Rholang only processes can be logically connected.
+
+## Wildcards vs Variables
+There is an important, somewhat subtle difference between wildcards and variables in Rholang. A wildcard can be placed anywhere that a *free* variable can be placed, meaning that wildcards can be part of patterns but cannot be part of processes. For example, we cannot write
+
+    for( _ <- @Nil){ _!(Nil) }
+
+due to the wildcard being in the body of the listen, even though it is correct to write
+
+    for( x <- @Nil){ x!(Nil) }
+
+It is, however, perfectly acceptable to write
+
+    for( _ <- @Nil){ Nil }
+
+That is the first distinction. The second distinction is that wildcards can bind to statements containing free variables, while variables cannot. For example, the pattern
+
+    new x in { p }
+
+will match to *strictly fewer* patterns than the pattern
+
+    new x in { _ }
+
+`new x in { p }` will only match to a process which does not ever use the new channel in the body, such as `new x in { Nil }`. Here, `p` binds to `Nil`. It cannot match to something like `new x in { x!(Nil) }`, because `p` would have to match with `x!(Nil)`, which by itself has the free variable `x`. On the other hand, `new x in { _ }` will match with `new x in { x!(Nil) }`, throwing away `x!(Nil)`. Since `_` doesn't actually bind to anything, we are allowed to give it things with free variables. We note here that the rules for patterns within patterns still apply, meaning that a wildcard cannot be used to match to a piece of a pattern, unless that pattern itself uses a wildcard in precisely the same way.
```

Comment:
This whole paragraph is italicized for seemingly no reason

Reply:
This is strange, I'm not sure how code gets italicized? It's not italicized anywhere else I see this code ...

Reply:
Yeah, it seems to start from `_!(Nil) }`. I think we can just ignore it for now.

Reply:
Weird.